### ISSUE TITLE
Pyright can’t find a submodule when I import the root package

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#3666]
### Discussed in https://github.com/microsoft/pyright/discussions/3663

<div type='discussions-op-text'>

<sup>Originally posted by **Pyprohly** July  6, 2022</sup>
```python
# import pkg
import pkg.util.foo
pkg.util.foo.Foo()
```

If I uncomment the first import in the preceding code, Pyright reports that the module `foo` does not exist. Should this happen?

Note I have an explicit import for module `util` in `pkg/__init__.py` which seems to be the cause of the issue.

```text
pkg
├── __init__.py
│   from . import util
│
└── util
    ├── __init__.py
    └── foo.py
        class Foo:
            pass
```
</div>

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.258, which I just published. It will also be included in the next release of pylance.

**Comment by ilrico** [comment]
I have a similar issue with python-dateutil, pyright latest (1.1.284)
```
import dateutil as du
import dateutil.relativedelta as du_relativedelta
du.relativedelta.relativedelta(month=1)  # ERROR: "relativedelta"is not a known member of module
du_relativedelta.relativedelta(month=1) # OK
```

**Comment by erictraut** [comment]
@ilrico, that is expected behavior. The `dateutil` type stub indicates that the `dateutil` module does not import (and re-export) the `relativedelta` submodule. That means you need to import the submodule explicitly (`import dateutil.relativedelta`) rather than relying on its parent module to load it.

It looks like this might be a bug in the stub because the implementation actually does import (and re-export) `relativedata` via an `__all__` reference. If you'd like, you can report it to the [typeshed project](https://github.com/python/typeshed/issues). All of the major Python type checkers, including pyright, get their stdlib type stubs from typeshed. The `dateutil.__init__.pyi` stub would need to add statements like `from . import relativedelta` to indicate to type checkers that the submodule is automatically loaded by its parent module.

**Comment by ilrico** [comment]
OK. thanks for this detailed explanation. I suspected it was pyright issue, since I had same with psycopg v3:  
```
import psycopg
psycopg.sql.SQL()  # ERROR: "sql" is not a known member of module
```
I'll create issues with typeshed.
